### ISSUE TITLE
Improve  OAuthConfig type

### ISSUE THREAD

**Issue by bzums** [zalando-incubator/authmosphere#110]
Currently there is one [OAuthConfig type](https://github.com/zalando-incubator/authmosphere/blob/master/src/types/OAuthConfig.ts#L1) for all supported grant types:
```typescript
type OAuthConfig = {
  credentialsDir: string;
  grantType: string; // (`AUTHORIZATION_CODE_GRANT` | `PASSWORD_CREDENTIALS_GRANT`)
  accessTokenEndpoint: string;
  scopes?: string[];
  redirect_uri?: string; // (required with `AUTHORIZATION_CODE_GRANT`)
  code?: string; // (required with `AUTHORIZATION_CODE_GRANT`)
  redirectUri?: string;
  refreshToken?: string;
  queryParams?: {};
};
```

This makes it difficult for the user to understand which properties are mandatory given a certain `grantType`. It would be more helpful to the user if `OAuthConfig` would be a [union type](https://www.typescriptlang.org/docs/handbook/advanced-types.html) of a number of explicit types, one for each supported grant type:

```typescript
type CredentialsGrantConfig = {
  credentialsDir: string;
  grantType: 'client_credentials';
  accessTokenEndpoint: string;
  scopes?: string[];
  queryParams?: {};
};

type AuthorizationCodeGrantConfig = {
  ...
};

type PasswordCredentialsGrantConfig = {
  ...
};

type RefreshGrantConfig = {
  ...
};

type OAuthConfig =
  CredentialsGrantConfig |
  AuthorizationCodeGrantConfig |
  PasswordCredentialsGrantConfig |
  RefreshGrantConfig;
```

This will lead to a number of refactorings and maybe introduces a breaking change, however the type support for the config is much more explicit.

The validation of the OAuthConfig will then be updated accordingly, see See the current validation of OAuthConfig: https://github.com/zalando-incubator/authmosphere/blob/master/src/oauth-tooling.ts#L185 resp. https://github.com/zalando-incubator/authmosphere/blob/master/src/utils.ts#L143

**Comment by awaltert** [comment]
@bzums Looks interesting, I have some questions regarding the requirements.

Do the different config types share some properties? e.g. `credentialsDir`, ...
Without looking at the implementation I asked myself, if the `queryParams` property really is typeof  `{}` or is it more like a `Dictionary<string, sting>` or `{[index: string]: string}`? 

What is the purpose of the two different `redirect_uri` props? Which kind belongs to which types?

I'd be happy to help with this issue 😄

**Comment by bzums** [comment]
@awaltert You are right, `queryParams` is rather `{[index: string]: string}` :)

The two different redirect uris are definitely  an error ^^ good you spotted it! As far as I see it `redirectUri` is correct and `redirect_uri` should be removed!

**Comment by awaltert** [comment]
@bzums Cool, thanks for the fast response. 👍  I'll adapt your feedback on `queryParams` and `redirectUri`.

Can you please elaborate on the question, which properties are common to the types and which are dedicated to which config type?

**Comment by ISO50** [comment]
closed by #115
